packages/controllers/src/snaps/Timer.ts

```diff
@@ -0,0 +1,57 @@
+export class Timer {
+  private _timeout?: NodeJS.Timeout | number;
+
+  private _start?: number;
+
+  private _remaining: number;
+
+  private _callback?: () => void;
+
+  constructor(ms: number) {
+    this._remaining = ms;
+  }
+
+  isStarted() {
+    return Boolean(this._timeout && this._start);
+  }
+
+  isFinished() {
+    return this._remaining <= 0;
+  }
+
+  cancel() {
+    this._remaining = 0;
+    clearTimeout(this._timeout as any);
+  }
+
+  pause() {
+    if (!this.isStarted()) {
+      return;
+    }
+    clearTimeout(this._timeout as any);
+    this._timeout = undefined;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this._remaining -= Date.now() - this._start!;
+  }
+
+  start(callback: () => void) {
```

Comment:
starting again while the Timer is started should throw an error, because there's something wrong with our assumptions in that case. This is specially important if the callback is different.